sdk/lib/rtl/rangelist.c

```diff
@@ -47,237 +509,675 @@ typedef struct _RTL_RANGE_ENTRY
+
+START:
+
+    RtlEntry = RtlpEntryFromLink(RangeList->ListHead.Flink);
+    NextRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink);
+
+    while (&RtlEntry->ListEntry != &RangeList->ListHead)
+    {
+        if (RtlEntry->PrivateFlags & RTLP_ENTRY_IS_MERGED)
         {
-            Current = CONTAINING_RECORD(Entry, RTL_RANGE_ENTRY, Entry);
-            if (Current->Range.Start > RangeEntry->Range.End)
+            MergedRtlEntry = RtlpEntryFromLink(RtlEntry->Merged.ListHead.Flink);
+            MergedNextEntry = RtlpEntryFromLink(MergedRtlEntry->ListEntry.Flink);
+
+            while (&MergedRtlEntry->ListEntry != &RtlEntry->Merged.ListHead)
             {
-                /* Insert before current */
-                DPRINT("Insert before current\n");
-                InsertTailList(&Current->Entry,
-                               &RangeEntry->Entry);
-
-                RangeList->Count++;
-                RangeList->Stamp++;
-                return STATUS_SUCCESS;
+                if (MergedRtlEntry->Allocated.Owner == Owner)
+                {
+                    DPRINT("RtlDeleteOwnersRanges: Deleting merged range [%I64X-%I64X]\n", MergedRtlEntry->Start, MergedRtlEntry->End);
+
+                    Status = RtlpDeleteFromMergedRange(RtlEntry, MergedRtlEntry);
+                    if (!NT_SUCCESS(Status))
+                    {
+                        DPRINT("RtlDeleteOwnersRanges: Status %X\n", Status);
+                        return Status;
+                    }
+
+                    RangeList->Count--;
+                    RangeList->Stamp++;
+
+                    goto START;
+                }
+
+                MergedRtlEntry = MergedNextEntry;
+                MergedNextEntry = RtlpEntryFromLink(MergedRtlEntry->ListEntry.Flink);
             }
+        }
+        else if (RtlEntry->Allocated.Owner == Owner)
+        {
+            DPRINT("RtlDeleteOwnersRanges: Deleting range [%I64X-%I64X]\n", RtlEntry->Start, RtlEntry->End);
+
+            RemoveEntryList(&RtlEntry->ListEntry);
+            //ExFreeToPagedLookasideList(&RtlpRangeListEntryLookasideList, RtlEntry);
+            RtlpFreeMemory(RtlEntry, 'elRR');
 
-            //Previous = Current;
-            Entry = Entry->Flink;
+            RangeList->Count--;
+            RangeList->Stamp++;
+
+            Status = STATUS_SUCCESS;
         }
 
-        DPRINT("Insert tail\n");
-        InsertTailList(&RangeList->ListHead,
-                       &RangeEntry->Entry);
-        RangeList->Count++;
-        RangeList->Stamp++;
-        return STATUS_SUCCESS;
+        RtlEntry = NextRtlEntry;
+        NextRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink);
     }
 
-    RtlpFreeMemory(RangeEntry, 0);
+    DPRINT("RtlDeleteOwnersRanges: return Status %X\n", Status);
 
-    return STATUS_UNSUCCESSFUL;
+    return Status;
 }
 
+/**********************************************************************
+ * NAME							EXPORTED
```

Comment:
As you are changing the code, please format the header according to our current [API Documentation](https://reactos.org/wiki/Documentation_Guidelines#API_Documentation) style (here and other functions)

Reply:
Done